### ISSUE TITLE
SDL2DesktopWindow::pollMouse() makes incorrect assumptions in regards to the output of SDL_GetGlobalMouseState

### ISSUE THREAD

**Issue by sp1ritCS** [ppy/osu-framework#5398]
The `pollMouse()` method gets called each time there isn't a cursor inside the window. This results in the function being called once each time my pen goes out of range of the tablet.

The issue arises, that according to [the docs](https://wiki.libsdl.org/SDL_GetGlobalMouseState) `SDL_GetGlobalMouseState` reports the global pointer position
> but the coordinates will be reported relative to the top-left of the desktop

now, the window position from `WindowMode.Value == Configuration.WindowMode.Windowed ? Position : windowDisplayBounds.Location;` is relative to the whole monitor coordinate system instead of the desktop the pointer is currently in.

This means, that unless osu!Lazer is running on the desktop with coordinate `(0|0)`, the output of GetGlobalMouseState (after the pen has left the tablet area above Lazer) could look like this: `(975|477)`, while the window position will be `(3200|0)`. This results in Lazer keeping the correct Y-coordinate of the cursor (in this case. not if the monitor displaying it is below the monitor containing the coordinate root), but improperly setting the X-coordinate of the pointer to 0.

I fail to understand the use of the `pollMouse()` method, and at least in my case, commenting out the call to `pollMouse()` makes it work as intended, with no other issues detected.

---

This is on Wayland.

**Comment by Susko3** [comment]
`SDL_GetGlobalMouseState` works fine for me in multi-monitor setup. Mouse positions are reported relative to the 0,0 of the primary monitor.

> is relative to the whole monitor coordinate system instead of the desktop the pointer is currently in.

I think you're misreading what `desktop` means here. It's contrasting the _window_ coordinate system. `desktop` is equivalent to your understanding of monitor coordinate system.

Can you reproduce the problem when using a mouse? If you move your mouse ever so slightly outside the window, the osu! cursor should still be visible and should follow the OS cursor. Test on both the primary and secondary monitor.

**Comment by sp1ritCS** [comment]
I did some more testing and realized, that on Wayland `SDL_GetGlobalMouseState` returns coordinates relative to the top left corner of the Lazer window due on Wayland. Kinda makes sense when you think about it (there is no way to get the pointer position outside the current window).

**Comment by Susko3** [comment]
`SDL_GetGlobalMouseState` is [not implemented](https://github.com/libsdl-org/SDL/blob/97a5e744497ff7cc93edc5119d67cad3ee86dd57/src/video/wayland/SDL_waylandmouse.c#L551-L557) on wayland, so SDL falls back to querying the [window mouse position](https://github.com/libsdl-org/SDL/blob/92b3c53c92971e685254fd89f89ce6bde8cea60e/src/events/SDL_mouse.c#L992-L994).

This is something to report upstream at https://github.com/libsdl-org/SDL/issues.

**Comment by sp1ritCS** [comment]
Yes, and by nature of the Wayland protocol it cannot be implemented, as any application can never do any actions outside their scope (so the window it has).
This will need to be fixed in osu-framrwork (simply not running the move cursor with pointer outside windows code on wayland seems reasonable, as it inherently doesn't work.)

**Comment by Susko3** [comment]
Framework uses [`SDL_WarpMouseInWindow`](https://github.com/ppy/osu-framework/blob/a28777eae4adbfe0bf0ffb1ca60b103c0a80de1d/osu.Framework/Platform/SDL2DesktopWindow.cs#L1489-L1494) to warp the mouse and this is a [no-op on wayland](https://github.com/libsdl-org/SDL/blob/97a5e744497ff7cc93edc5119d67cad3ee86dd57/src/video/wayland/SDL_waylandmouse.c#L522-L532).

It's really strange that the cursor is warping in the first place. Even stranger that commenting out `pollMouse()` fixes it.

**Comment by sp1ritCS** [comment]
> It's really strange that the cursor is warping in the first place.

maybe it's because it doesn't query the pointer position of the tablet anymore at this point and instead receives the pointer position of the mouse? (the compositor doesn't unify the tablet and mouse pointer into a single cursor)

> Even stranger that commenting out pollMouse() fixes it.

does it? it seems like it makes sense for me.

**Comment by Susko3** [comment]
Could you make a screen recording with both your monitors visible?

**Comment by sp1ritCS** [comment]
Top left monitor starts at (0|0), top right monitor starts at (1280|0) and bottom right monitor starts at (3200|0).
Bottom left is just the osu! std{out,err}.
https://user-images.githubusercontent.com/31540351/189478972-4284399e-0412-4e0f-a2f3-158a1677f490.mp4

**Comment by Susko3** [comment]
I think the main problem here is `CursorInWindow` becoming `false`.

This needs more investigation. Could OS tablet handling be interfering here?

The cursor jumping (while caused by lack of support in wayland) wouldn't happen if the `CursorInWindow` value was correct.

**Comment by sp1ritCS** [comment]
> Could OS tablet handling be interfering here?

the OS doesn't do much with the tablet, if the pointer is inside any wayland surface. Wayland exposes tablet events separately from cursor events using the [`tablet-unstable-v2` API](https://wayland.app/protocols/tablet-unstable-v2). Maybe the [code I've written](https://github.com/libsdl-org/SDL/pull/5424) in SDL to expose tablet events as mouse events isn't handling multiple devices correctly, but I wouldn't know where to look in regards to `CursorInWindow`?
